tempesta_fw/cache.c

```diff
@@ -1173,58 +1455,108 @@ tfw_cache_copy_resp(TfwCacheEntry *ce, TfwHttpResp *resp, size_t tot_len)
 		tot_len -= n;
 		ce->key_len += n;
 	}
+
 	/* Request method is a part of the cache record key. */
 	ce->method = req->method;
 
+	/* Write ':status' pseudo-header. */
 	ce->status = TDB_OFF(db->hdr, p);
-	if ((n = tfw_cache_strcpy_eol(&p, &trec, s_line, &tot_len, 1)) < 0) {
-		T_ERR("Cache: cannot copy HTTP status line\n");
-		return -ENOMEM;
+	status_idx = tfw_h2_pseudo_index(resp->status);
+	if (status_idx) {
+		TfwHPackInt hp_idx;
+		TfwStr str = {};
+
+		write_int(status_idx, 0x7F, 0x80, &hp_idx);
+		str.data = hp_idx.buf;
+		str.len = hp_idx.sz;
+
+		if (tfw_cache_h2_copy_str(&ce->status_len, &p, &trec, &str,
+					  &tot_len))
+			return -ENOMEM;
+	} else {
+		char buf[H2_STAT_VAL_LEN];
+		TfwStr str = {
+			.data = buf,
+			.len = H2_STAT_VAL_LEN
+		};
+
+		/*
+		 * If the ':status' pseudo-header is not fully indexed, set
+		 * the default static index (8) just for the name.
+		 */
+		if (tfw_cache_h2_copy_int(&ce->status_len, 8, 0xF, &p, &trec,
+					  &tot_len)
+		    || tfw_cache_h2_copy_int(&ce->status_len, H2_STAT_VAL_LEN,
+					     0x7f, &p, &trec, &tot_len))
+			return -ENOMEM;
+
+		if (!tfw_ultoa(resp->status, str.data, H2_STAT_VAL_LEN))
+			return -E2BIG;
+
+		if (tfw_cache_h2_copy_str(&ce->status_len, &p, &trec, &str,
+					  &tot_len))
+			return -ENOMEM;
 	}
-	ce->status_len += n;
+
+	if (tfw_cache_h2_copy_str(&ce->rph_len, &p, &trec, rph, &tot_len))
+		return -ENOMEM;
 
 	ce->hdrs = TDB_OFF(db->hdr, p);
 	ce->hdr_len = 0;
 	ce->hdr_num = resp->h_tbl->off;
 	FOR_EACH_HDR_FIELD_FROM(field, end1, resp, TFW_HTTP_HDR_REGULAR) {
-		bool hdr_304 = false;
-
-		/* Skip hop-by-hop headers. */
-		if (!(field->flags & TFW_STR_HBH_HDR)) {
-			h = field;
-		} else if (field - resp->h_tbl->tbl < TFW_HTTP_HDR_RAW) {
-			h = &empty;
-		} else {
+		int hid = field - resp->h_tbl->tbl;
+		/*
+		 * Skip hop-by-hop headers. Also skip 'Server' header (with
+		 * possible duplicates), since we will substitute it with our
+		 * version of this header.
+		 */
+		if ((field->flags & TFW_STR_HBH_HDR)
+		    || hid == TFW_HTTP_HDR_SERVER
```

Comment:
Let's skip `Set-Cookie` header. Actually backend should protect responses with private cookies from caching, but I'll be good not to spread the same cookie among all clients if it mistakenly leaks. What do you think?

Reply:
As a result of discussion in the chat, it was decided to postpone this change and additionally investigate it.

Reply:
Made some research and moved the requirement to #1075 which is all about cookies and security.